torch/distributed/_functional_collectives.py

```diff
@@ -0,0 +1,196 @@
+from typing import Any, Tuple, Union, List, cast
+
+import torch
+from torch.autograd import Function
+from torch.utils.weak import WeakIdKeyDictionary
+import torch.distributed as dist
+
+from torch._C import _disabled_torch_function_impl
+from torch.utils._pytree import tree_map
+
+import torch.distributed.distributed_c10d as c10d
+import torch.distributed._tensor as dt
+
+"""
+New traceable, functional collectives.
+  compiler: trace these ops with plain-old-data schemas, then choose how to lower them.
+  eager: execute these 'functional' ops which in eager return AsyncCollectiveTensor subclasses,
+         automatically calling .wait() on underlying/hidden async 'work' obj only when fed to
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* How can we make these ops work in eager without manually enabling python_dispatacher mode?
+
+"""
+
+"""
+FIXME for this to work correctly we need to change Work to internally hold no reference to the tensor.
+"""
```

Comment:
please clean these up before landing.

Reply:
Oh I was referring to clean up the docs and comments, but addressing the real issues. E.g., #FIXME was supposed to be in comments not string.

Reply:
oh, lol. Fixed then :)